components/brave_wallet/browser/zcash/rust/orchard_block_decoder_impl.cc

```diff
@@ -20,56 +21,61 @@ OrchardBlockDecoderImpl::OrchardBlockDecoderImpl(const OrchardFullViewKey& fvk)
+    const ::brave_wallet::OrchardTreeState& tree_state,
+    const std::vector<::brave_wallet::zcash::mojom::CompactBlockPtr>& blocks) {
+  ::rust::Vec<orchard::OrchardCompactAction> orchard_actions;
+  for (const auto& block : blocks) {
+    bool block_has_orchard_action = false;
+    for (const auto& tx : block->vtx) {
+      for (const auto& orchard_action : tx->orchard_actions) {
+        block_has_orchard_action = true;
+        orchard::OrchardCompactAction orchard_compact_action;
 
-      base::ranges::copy(orchard_action->nullifier,
-                         orchard_compact_action.nullifier.begin());
-      base::ranges::copy(orchard_action->cmx,
-                         orchard_compact_action.cmx.begin());
-      base::ranges::copy(orchard_action->ephemeral_key,
-                         orchard_compact_action.ephemeral_key.begin());
-      base::ranges::copy(orchard_action->ciphertext,
-                         orchard_compact_action.enc_cipher_text.begin());
+        if (orchard_action->nullifier.size() != kOrchardNullifierSize ||
+            orchard_action->cmx.size() != kOrchardCmxSize ||
+            orchard_action->ephemeral_key.size() != kOrchardEphemeralKeySize ||
+            orchard_action->ciphertext.size() != kOrchardCipherTextSize) {
+          return nullptr;
+        }
 
-      orchard_actions.emplace_back(std::move(orchard_compact_action));
-    }
+        orchard_compact_action.block_id = block->height;
+        orchard_compact_action.is_block_last_action = false;
+        base::ranges::copy(orchard_action->nullifier,
+                           orchard_compact_action.nullifier.begin());
+        base::ranges::copy(orchard_action->cmx,
+                           orchard_compact_action.cmx.begin());
+        base::ranges::copy(orchard_action->ephemeral_key,
+                           orchard_compact_action.ephemeral_key.begin());
+        base::ranges::copy(orchard_action->ciphertext,
+                           orchard_compact_action.enc_cipher_text.begin());
 
-    ::rust::Box<::brave_wallet::orchard::BatchOrchardDecodeBundleResult>
-        decode_result = ::brave_wallet::orchard::batch_decode(
-            full_view_key_, std::move(orchard_actions));
-
-    if (decode_result->is_ok()) {
-      ::rust::Box<::brave_wallet::orchard::BatchOrchardDecodeBundle>
-          result_bundle = decode_result->unwrap();
-      for (size_t i = 0; i < result_bundle->size(); i++) {
-        result.emplace_back(
-            OrchardNote({{},
-                         block->height,
-                         result_bundle->note_nullifier(full_view_key_, i),
-                         result_bundle->note_value(i),
-                         0,
-                         {},
-                         {}}));
+        orchard_actions.push_back(std::move(orchard_compact_action));
       }
-    } else {
-      return std::nullopt;
     }
+    if (block_has_orchard_action) {
```

Comment:
let's just check for `!orchard_actions.empty()`

Reply:
orchard_actions contains orchard actions across several blocks, so we need to mark last one for each block